### PR TITLE
rustdoc: Add more semantic information to impl IDs

### PR DESCRIPTION
**Pull request by pierwill**
Instead of generating `#impl`, `#impl-1`, etc., generate IDs like `#impl-Foo<M>`.

Part of https://github.com/rust-lang/rust/issues/92052.

### CODE REVIEW

**Comment by camelid** [comment]
Can you update the tests?

@rustbot author

**Comment by pierwill** [comment]
> Can you update the tests?
> 
> @rustbot author

Should `x test src/test/rustdoc --bless` work for this? I am getting many errors that don't seem related...

**Comment by camelid** [comment]
No, you'll need to update the `@has` comments manually.

**Comment by pierwill** [comment]
I'm getting panics on at least one test that has no `@has`:

```
---- [rustdoc] rustdoc/async-move-doctest.rs stdout ----

error: rustdoc failed!
status: exit status: 101
command: "/Users/user/repos/rust/build/x86_64-apple-darwin/stage1/bin/rustdoc" "-L" "/Users/user/repos/rust/build/x86_64-apple-darwin/stage1/lib/rustlib/x86_64-apple-darwin/lib" "-L" "/Users/user/repos/rust/build/x86_64-apple-darwin/test/rustdoc/async-move-doctest/auxiliary" "-o" "/Users/user/repos/rust/build/x86_64-apple-darwin/test/rustdoc/async-move-doctest" "--deny" "warnings" "/Users/user/repos/rust/src/test/rustdoc/async-move-doctest.rs" "--test" "--edition=2018"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'rustc' panicked at 'SESSION_GLOBALS should never be overwritten! Use another thread if you need another SessionGlobals', /Users/user/repos/rust/compiler/rustc_span/src/lib.rs:103:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-dev running on x86_64-apple-darwin

query stack during panic:
end of query stack

------------------------------------------
```

**Comment by camelid** [comment]
Hmm, try `rm -r build/<your-machine-arch>/stage*` and then re-running tests. Might just be corruption in your build directory.

**Comment by pierwill** [comment]
Still getting nasty panics. 😬

**Comment by camelid** [comment]
Weird, what's your `config.toml`?

**Comment by pierwill** [comment]
```
# Includes one of the default files in src/bootstrap/defaults
profile = "compiler"
changelog-seen = 2

[rust]
incremental = true

# deny-warnings = false

debug = true
debug-assertions = false
debuginfo-level = 1
debuginfo-level-rustc = 1
debuginfo-level-std = 1

run-dsymutil = false

parallel-compiler = true

[build]
compiler-docs = true



# set to false for src/ci/docker runs
# submodules = false
```

**Comment by camelid** [comment]
Oh, `parallel-compiler` doesn't work properly. You should disable it. I think that's what's causing the panics.

**Comment by pierwill** [comment]
> Oh, `parallel-compiler` doesn't work properly. You should disable it. I think that's what's causing the panics.
> 
> I'm curious, why did you enable it? Did you see it recommended somewhere?

I was working on something with parallelism and incr. comp...

**Comment by pierwill** [comment]
Having trouble with some remaining test failures:

```
failures:
    [rustdoc] rustdoc/blanket-reexport-item.rs
    [rustdoc] rustdoc/const-generics/const-impl.rs
    [rustdoc] rustdoc/generic-impl.rs
    [rustdoc] rustdoc/primitive/primitive-generic-impl.rs
    [rustdoc] rustdoc/src-links-auto-impls.rs
```

**Comment by pierwill** [comment]
Possibly this change should have affected [this line](https://github.com/rust-lang/rust/pull/92745/files#diff-92a2beaff0ae8c02e2f54cae107b0ccbb8d2795abbb45786210573d0e3074d38R27), but it has not:

```
// @has foo/struct.VSet.html '//div[@id="impl-1"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
```

**Comment by camelid** [comment]
Could you also update the GUI tests that are failing in CI? The errors that CI shows should be enough to fix them.

**Comment by camelid** [comment]
> Possibly this change should have affected [this line](https://github.com/rust-lang/rust/pull/92745/files#diff-92a2beaff0ae8c02e2f54cae107b0ccbb8d2795abbb45786210573d0e3074d38R27), but it has not:
> 
> ```
> // @has foo/struct.VSet.html '//div[@id="impl-1"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
> ```

Indeed, that is strange.

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -1674,19 +1670,12 @@ pub(crate) fn render_impl_summary(
...
     let id = cx.derive_id(match i.inner_impl().trait_ {
-        Some(ref t) => {
-            if is_on_foreign_type {
-                get_id_for_impl_on_foreign_type(&i.inner_impl().for_, t, cx)
-            } else {
-                format!("impl-{}", small_url_encode(format!("{:#}", t.print(cx))))
-            }
-        }
+        Some(ref t) => get_id_for_impl(&i.inner_impl().for_, t, cx),
         None => "impl".to_string(),
```
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:1679]: Oh, that explains it! You need to update this line as well. You'll need to make `get_id_for_impl`'s trait argument be an Option.

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -2194,11 +2183,7 @@ fn sidebar_struct(cx: &Context<'_>, buf: &mut Buffer, it: &clean::Item, s: &clea
...
 }
 
-fn get_id_for_impl_on_foreign_type(
-    for_: &clean::Type,
-    trait_: &clean::Path,
-    cx: &Context<'_>,
-) -> String {
+fn get_id_for_impl(for_: &clean::Type, trait_: &clean::Path, cx: &Context<'_>) -> String {
     small_url_encode(format!("impl-{:#}-for-{:#}", trait_.print(cx), for_.print(cx)))
 }
```
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:2188]:
```suggestion
fn get_id_for_impl(for_: &clean::Type, trait_: Option<&clean::Path>, cx: &Context<'_>) -> String {
    match trait_ {
        Some(t) => small_url_encode(format!("impl-{:#}-for-{:#}", t.print(cx), for_.print(cx))),
        None => small_url_encode(format!("impl-{:#}", for_.print(cx))),
    }
}
```

**Comment by pierwill** [comment]
This test is still passing when it should fail: https://github.com/rust-lang/rust/blob/33800af5c55c8eaaeb649715bc85fccdc362c6f8/src/test/rustdoc/const-generics/const-impl.rs#L27.

**Comment by pierwill** [comment]
Can we use `GenericParamDefKind::Const` somehow in `get_id_for_impl`? 🤔

File: src/librustdoc/html/render/mod.rs [hunk 3]
```diff
@@ -1674,19 +1670,12 @@ pub(crate) fn render_impl_summary(
...
     let id = cx.derive_id(match i.inner_impl().trait_ {
-        Some(ref t) => {
-            if is_on_foreign_type {
-                get_id_for_impl_on_foreign_type(&i.inner_impl().for_, t, cx)
-            } else {
-                format!("impl-{}", small_url_encode(format!("{:#}", t.print(cx))))
-            }
-        }
+        Some(ref t) => get_id_for_impl(&i.inner_impl().for_, Some(t), cx),
         None => "impl".to_string(),
```
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:1679]: > This test is still passing when it should fail:

@pierwill you forgot to update this line ;)

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:1679]: You can just get rid of the whole match and pass `i.inner_impl().trait_.as_ref()` directly as the argument to `get_id_for_impl`.

Comment (pierwill) [reply src/librustdoc/html/render/mod.rs:1679]: I don't get it 😫 lol

Comment (camelid) [reply src/librustdoc/html/render/mod.rs:1679]: Current:

```rust
    let id = cx.derive_id(match i.inner_impl().trait_ {
        Some(ref t) => get_id_for_impl(&i.inner_impl().for_, Some(t), cx),
        None => "impl".to_string(),
    });
```

What it needs to be:

```rust
    let id = cx.derive_id(
        get_id_for_impl(&i.inner_impl().for_, i.inner_impl().trait_.as_ref(), cx),
    });
```

File: src/test/rustdoc/const-generics/const-impl.rs [hunk 4]
```diff
@@ -33,7 +33,7 @@ impl<T> VSet<T, { Order::Unsorted }> {
 
 pub struct Escape<const S: &'static str>;
 
-// @has foo/struct.Escape.html '//div[@id="impl"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
+// @has foo/struct.Escape.html '//div[@id="impl-Escape%3Cr%23%22%3Cscript%3Ealert%28%22Escape%22%29%253B%3C%2Fscript%3E%22%23%3E"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
 impl Escape<r#"<script>alert("Escape");</script>"#> {
```
Comment (pierwill) [inline src/test/rustdoc/const-generics/const-impl.rs:37]: These are copied from the resulting HTML files and then URL encoded... still not passing...

Comment (camelid) [reply src/test/rustdoc/const-generics/const-impl.rs:37]: Hmm, what do you mean "and then URL encoded"? I think just copying them straight from the HTML should work.

Comment (pierwill) [reply src/test/rustdoc/const-generics/const-impl.rs:37]: I thought special characters need to be "percent encoded" here. Copying this string from the URL doesn't work.

Comment (pierwill) [reply src/test/rustdoc/const-generics/const-impl.rs:37]: https://github.com/rust-lang/rust/pull/92745/commits/515eba72468aa30b27eb40bd17f5b550753d9650

Comment (jsha) [reply src/test/rustdoc/const-generics/const-impl.rs:37]: This diff gets the test to pass for me.

```diff
diff --git a/src/test/rustdoc/const-generics/const-impl.rs b/src/test/rustdoc/const-generics/const-impl.rs
index bd961fa6434..69a7cee71f3 100644
--- a/src/test/rustdoc/const-generics/const-impl.rs
+++ b/src/test/rustdoc/const-generics/const-impl.rs
@@ -1,7 +1,5 @@
 #![allow(incomplete_features)]
-
 #![feature(adt_const_params)]
-
 #![crate_name = "foo"]
 
 #[derive(PartialEq, Eq)]
@@ -17,14 +15,14 @@ pub struct VSet<T, const ORDER: Order> {
     inner: Vec<T>,
 }
 
-// @has foo/struct.VSet.html '//div[@id="impl-VSet<T%2C%20{%20Order%3A%3ASorted%20}>"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Sorted }>'
+// @has foo/struct.VSet.html '//div[@id="impl-VSet%3CT%2C%20{%20Order%3A%3AUnsorted%20}%3E"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
 impl<T> VSet<T, { Order::Sorted }> {
     pub fn new() -> Self {
         Self { inner: Vec::new() }
     }
 }
 
-// @has foo/struct.VSet.html '//div[@id="impl-VSet<T%2C%20{%20Order%3A%3AUnsorted%20}>"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
+// @has foo/struct.VSet.html '//div[@id="impl-VSet%3CT%2C%20{%20Order%3A%3AUnsorted%20}%3E"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
 impl<T> VSet<T, { Order::Unsorted }> {
     pub fn new() -> Self {
         Self { inner: Vec::new() }
@@ -33,7 +31,7 @@ pub fn new() -> Self {
 
 pub struct Escape<const S: &'static str>;
 
-// @has foo/struct.Escape.html '//div[@id="impl-Escape<r#"<script>alert("Escape")%3B</script>"#>"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
+// has foo/struct.Escape.html '//div[@id="impl-Escape<r#"<script>alert("Escape")%3B</script>"#>"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
 impl Escape<r#"<script>alert("Escape");</script>"#> {
     pub fn f() {}
 }
```

Comment (jsha) [reply src/test/rustdoc/const-generics/const-impl.rs:37]: My process was this: I went into the HTML and copied the ID, making sure to copy it from the `id=` field, not the `href=` field. Also, I got it from the raw HTML, not the Dev Tools console.

Also, when debugging these XPath tests it's useful to know that the text content at the end is a substring match. So if you're getting match failures, you can set the text content to `''` and see if you get a match, then gradually add back text and see if it still passes.

File: src/test/rustdoc/const-generics/const-impl.rs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+// ignore-tidy-linelength
```
Comment (camelid) [inline src/test/rustdoc/const-generics/const-impl.rs:1]: This isn't needed, it's implied in tests.

File: src/test/rustdoc/const-generics/const-impl.rs [hunk 6]
```diff
@@ -33,7 +35,7 @@ impl<T> VSet<T, { Order::Unsorted }> {
 
 pub struct Escape<const S: &'static str>;
 
-// @has foo/struct.Escape.html '//div[@id="impl"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
+// has foo/struct.Escape.html '//div[@id="impl-Escape<r#"<script>alert("Escape")%3B</script>"#>"]/h3[@class="code-header in-band"]' 'impl Escape<r#"<script>alert("Escape");</script>"#>'
```
Comment (camelid) [inline src/test/rustdoc/const-generics/const-impl.rs:38]: The removal of the `@` means this test doesn't do anything. I think it was a typo on @jsha's part.

File: src/test/rustdoc/const-generics/const-impl.rs [hunk 7]
```diff
@@ -11,20 +13,20 @@ pub enum Order {
...
-// @has foo/struct.VSet.html '//div[@id="impl-Send"]/h3[@class="code-header in-band"]' 'impl<T, const ORDER: Order> Send for VSet<T, ORDER>'
-// @has foo/struct.VSet.html '//div[@id="impl-Sync"]/h3[@class="code-header in-band"]' 'impl<T, const ORDER: Order> Sync for VSet<T, ORDER>'
+// @has foo/struct.VSet.html '//div[@id="impl-Send-for-VSet%3CT%2C%20ORDER%3E"]/h3[@class="code-header in-band"]' 'impl<T, const ORDER: Order> Send for VSet<T, ORDER>'
+// @has foo/struct.VSet.html '//div[@id="impl-Sync-for-VSet%3CT%2C%20ORDER%3E"]/h3[@class="code-header in-band"]' 'impl<T, const ORDER: Order> Sync for VSet<T, ORDER>'
 pub struct VSet<T, const ORDER: Order> {
     inner: Vec<T>,
 }
 
-// @has foo/struct.VSet.html '//div[@id="impl"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Sorted }>'
+// @has foo/struct.VSet.html '//div[@id="impl-VSet%3CT%2C%20{%20Order%3A%3AUnsorted%20}%3E"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Unsorted }>'
```
Comment (camelid) [inline src/test/rustdoc/const-generics/const-impl.rs:22]: This test is wrong:

```suggestion
// @has foo/struct.VSet.html '//div[@id="impl-VSet%3CT%2C%20{%20Order%3A%3ASorted%20}%3E"]/h3[@class="code-header in-band"]' 'impl<T> VSet<T, { Order::Sorted }>'
```

File: src/test/rustdoc/const-generics/const-impl.rs (hunk 6 above)
Comment (jsha) [reply src/test/rustdoc/const-generics/const-impl.rs:38]: That's correct, thanks for spotting that.

**Comment by camelid** [comment]
Ok, I think you're really close, this is the last test failure :)

```
39: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/struct.Foo.html '//div[@id="impl-Foo"]/h3[@class="code-header in-band"]' 'impl<const M: usize> Foo<M> where u8: Trait<M>'
50: @has check failed
	`XPATH PATTERN` did not match
	// @has foo/struct.Bar.html '//div[@id="impl-Bar"]/h3[@class="code-header in-band"]' 'impl<const M: usize> Bar<u8, M>'

Encountered 2 errors
```

**Comment by camelid** [comment]
One thing I just realized is that the new code doesn't disambiguate based on trait bounds. We can probably find a way to implement that later though.

File: src/test/rustdoc/const-generics/const-impl.rs (hunk 5 above)
Comment (pierwill) [reply src/test/rustdoc/const-generics/const-impl.rs:1]: This was because of the `@` typo.

**Comment by pierwill** [comment]
When I try to use the compiled rustdoc to build this example, I get an edition error:

```
❯  ./build/x86_64-apple-darwin/stage1/bin/rustdoc -o out src/test/rustdoc/const-generics/const-generics-docs.rs
error[E0670]: `async fn` is not permitted in Rust 2015
  --> /Users/user/repos/rust/src/test/rustdoc/const-generics/const-generics-docs.rs:67:5
   |
67 | pub async fn a_sink<const N: usize>(v: [u8; N]) -> impl Trait<N> {
   |     ^^^^^ to use `async fn`, switch to Rust 2018 or later
   |
   = help: set `edition = "2021"` in `Cargo.toml`
   = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0670]: `async fn` is not permitted in Rust 2015
  --> /Users/user/repos/rust/src/test/rustdoc/const-generics/const-generics-docs.rs:73:5
   |
73 | pub async fn b_sink<const N: usize>(_: impl Trait<N>) {}
   |     ^^^^^ to use `async fn`, switch to Rust 2018 or later
   |
   = help: set `edition = "2021"` in `Cargo.toml`
   = note: for more on editions, read https://doc.rust-lang.org/edition-guide

error[E0463]: can't find crate for `extern_crate`
 --> /Users/user/repos/rust/src/test/rustdoc/const-generics/const-generics-docs.rs:5:1
  |
5 | extern crate extern_crate;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: Compilation failed, aborting rustdoc

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0463, E0670.
For more information about an error, try `rustc --explain E0463`.
```

**Comment by camelid** [comment]
@pierwill Pass `--edition=2021`. The default edition is 2015 unless you specify otherwise (e.g., `Cargo.toml` does it for you).

**Comment by pierwill** [comment]
These tests are driving me crazy! 😆

**Comment by jsha** [comment]
Sorry to hear that! They are indeed frustrating. Want to ask on Zulip for some help sorting them out? I can also take a stab at it tomorrow.

File: src/test/rustdoc-gui/search-filter.goml [hunk 8]
```diff
@@ -14,4 +14,4 @@ press-key: "Enter"
 // Waiting for the search results to appear...
 wait-for: "#titles"
 // We check that there is no more "test_docs" appearing.
-assert-false: "#results .externcrate"
+// assert-false: "#results .externcrate"
```
Comment (pierwill) [inline src/test/rustdoc-gui/search-filter.goml:17]: Not sure how we want to handle this. @GuillaumeGomez

Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/search-filter.goml:17]: If the result appears, then it means the search filter is broken. Try to do on your browser what it's doing in this and you should see pretty quickly what's wrong. If you can't figure it out, I can take a look in the next days.

Please revert this change in any case.

Comment (pierwill) [reply src/test/rustdoc-gui/search-filter.goml]: Reverted.

**Comment by camelid** [comment]
Ok, now you're seeing all the broken links in the standard library 😅 

Before you go to the effort to update all of those, we should get confirmation from the team that this new anchor formatting looks good to them.

File: src/test/rustdoc-gui/search-filter.goml (hunk 8 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/search-filter.goml]: Nice!

**Comment by GuillaumeGomez** [comment]
I said it on zulip but doesn't hurt to say it here as well: I agree with this change (good luck updating all std URLs :wink: ).

**Comment by jsha** [comment]
I've put up a demo of this branch at https://rustdoc.crud.net/jsha/rustdoc-disamb-impls/std/fs/struct.File.html#impl-AsFd-for-File (@pierwill, if you'd like an account on this server so you can upload demos, PM me on Zulip! I'm @jsha there too).

I spotted the broken links issue. If you look in the sidebar for `AsFd`, and click on it, you're taken to https://rustdoc.crud.net/jsha/rustdoc-disamb-impls/std/fs/struct.File.html#impl-AsFd. But the new fragment is `#impl-AsFd-for-File`, so that's a broken link.

I _think_ this is the code you need to tweak: https://github.com/rust-lang/rust/blob/48e89b00caa94829a5f07e0f1ecb33bf37431244/src/librustdoc/html/render/mod.rs#L2012-L2033.

**Comment by pierwill** [comment]
I do still want to finish this. 🤓

**Comment by JohnCSimon** [comment]
@pierwill 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR BEFORE you force push to it, else you won't be able to reopen.
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**